src/main/java/org/springframework/data/couchbase/core/convert/MappingCouchbaseConverter.java

```diff
@@ -541,6 +541,7 @@ public void doWithPersistentProperty(final CouchbasePersistentProperty prop) {
 				generatedValueInfo = idProperty.findAnnotation(GeneratedValue.class);
 				String generatedId = generateId(generatedValueInfo, prefixes, suffixes, idAttributes);
 				target.setId(generatedId);
+				// this is not effective if id is Immutable, and accessor.setProperty() returns a new object in getBean()
```

Comment:
Id population should happen on the Template API level, see [MongoDB for reference](https://github.com/spring-projects/spring-data-mongodb/blob/2.0.x/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/MongoTemplate.java#L1054). Note, newer versions have the [Id population extracted](https://github.com/spring-projects/spring-data-mongodb/blob/master/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/MongoTemplate.java#L2717) because it happened in multiple places.

Reply:
After adding the update via CouchbaseTemplateSupport.applyUpdatedId() ( called by Reactive[Insert/Upsert]ByIdSupport.one(entity) which is called from the template), I left the update to the sourceObject here for two reasons - the first being that a number of unit tests depend on it (those can be fixed), the second is that if it is removed from here, it would appear to be a regression for anyone calling  `repository.save( sourceObject )`, and expecting sourceObject to contain the id. They should be  calling as  `savedObject = repository.save( sourceObject )`.   I could remove this now in hope that will result in fewer customers using it incorrectly, or wait to do it in the 4.1 release because it could be a breaking change for some.

My first idea was to have the Couchbase implementation of EntityWriter.write(source, target) return the updated object - but the signature returns void. (updating here and returning accessor.getBean() would have worked for that).

Reply:
Let's schedule any removal for the next version. We're in the RC phase with the GA release coming up Oct 28.

In any case in an async/reactive API usage scenario, it's hard to reason about when the model is being updated. In-place updates could have nasty side effects therefore we generally recommend using immutable types if you want to work with the saved/updated object instance that is returned by `save(…)`.